plugins/auto-sizes/hooks.php

```diff
@@ -139,79 +139,145 @@ function auto_sizes_get_width( string $layout_width, int $image_width ): string
 }
 
 /**
- * Filter the sizes attribute for images to improve the default calculation.
+ * Primes attachment into the cache with a single database query.
  *
- * @since 1.1.0
+ * @since n.e.x.t
  *
- * @param string                                                   $content      The block content about to be rendered.
- * @param array{ attrs?: array{ align?: string, width?: string } } $parsed_block The parsed block.
- * @return string The updated block content.
+ * @param string|mixed $content The HTML content.
+ * @return string The HTML content.
  */
-function auto_sizes_filter_image_tag( string $content, array $parsed_block ): string {
+function auto_sizes_prime_attachment_caches( $content ): string {
+	if ( ! is_string( $content ) ) {
+		return '';
+	}
+
 	$processor = new WP_HTML_Tag_Processor( $content );
-	$has_image = $processor->next_tag( array( 'tag_name' => 'img' ) );
 
-	// Only update the markup if an image is found.
-	if ( $has_image ) {
-		$processor->set_attribute( 'data-needs-sizes-update', true );
-		if ( isset( $parsed_block['attrs']['align'] ) ) {
-			$processor->set_attribute( 'data-align', $parsed_block['attrs']['align'] );
+	$images = array();
+	while ( $processor->next_tag( array( 'tag_name' => 'IMG' ) ) ) {
+		$class = $processor->get_attribute( 'class' );
+
+		if ( ! is_string( $class ) ) {
+			continue;
 		}
 
-		// Resize image width.
-		if ( isset( $parsed_block['attrs']['width'] ) ) {
-			$processor->set_attribute( 'data-resize-width', $parsed_block['attrs']['width'] );
+		if ( preg_match( '/(?:^|\s)wp-image-([1-9][0-9]*)(?:\s|$)/', $class, $class_id ) === 1 ) {
+			$attachment_id = (int) $class_id[1];
+			if ( $attachment_id > 0 ) {
+				$images[] = $attachment_id;
+			}
 		}
+	}
+
+	// Reduce the array to unique attachment IDs.
+	$attachment_ids = array_unique( $images );
 
-		$content = $processor->get_updated_html();
+	if ( count( $attachment_ids ) > 1 ) {
+		/*
+		 * Warm the object cache with post and meta information for all found
+		 * images to avoid making individual database calls.
+		 */
+		_prime_post_caches( $attachment_ids, false, true );
 	}
+
 	return $content;
 }
-add_filter( 'render_block_core/image', 'auto_sizes_filter_image_tag', 10, 2 );
-add_filter( 'render_block_core/cover', 'auto_sizes_filter_image_tag', 10, 2 );
+
+/*
+ * The priority 6 is used to ensure the new filter runs right after the "wp-image-$attachment_id" class
+ * is added to the img tag at priority 5, allowing modifications that rely on this class being in place.
+ */
+add_filter( 'the_content', 'auto_sizes_prime_attachment_caches', 6 );
```

Comment:
I don't believe the added comment is accurate. The `wp-image-$attachment_id` classname is not added on a filter, but is instead added to stored `img` markup saved to the database when an image block is authored.

Currently, WP adds `srcset` and `sizes` on `the_content` filter after the content is rendered. As part of that process, it primes the attachment data for all images in the content being filtered before looping through all of the images on the page and adding calculating `srcset` and `sizes` for each image. All of this logic happens in [wp_filter_content_tags()](https://github.com/WordPress/wordpress-develop/blob/61b9859728b31130274b31bde9c6d73ee915fa03/src/wp-includes/media.php#L1861). The cache priming is necessary to avoid making individual DB calls for each image being modified.

The approach explored in this PR moves the logic for adding the `sizes` value from post processing on `the_content`, to the block rendering process (i.e., by modifying the `sizes` value on the `render_block_core/image` filter). The reason the priming logic needs to run on an earlier priority is because it must run prior to the `do_blocks` process that [runs on 'the_content' filter at priority 9](https://github.com/WordPress/wordpress-develop/blob/61b9859728b31130274b31bde9c6d73ee915fa03/src/wp-includes/default-filters.php#L195).

Registering this hook on priority 8 still works as expected.

```suggestion
// This must run before 'do_blocks', which runs at priority 9.
add_filter( 'the_content', 'auto_sizes_prime_attachment_caches', 8 );
```